### ISSUE TITLE
[BUG]  Unable to install (again) after executing some system maintenance

### ISSUE THREAD

**Issue by jeff9finger-usda** [RedBearAK/toshy#289]
[toshy-services-log.txt](https://github.com/user-attachments/files/15776144/toshy-services-log.txt)
[toshy-services-status.txt](https://github.com/user-attachments/files/15776146/toshy-services-status.txt)
[toshy-out.txt](https://github.com/user-attachments/files/15776182/toshy-out.txt)
```
(EV) Toshy env module sees this environment:
                 DISTRO_ID       = 'rhel'
                 DISTRO_VER      = '8.10'
                 VARIANT_ID      = 'notfound'
                 SESSION_TYPE    = 'x11'
                 DESKTOP_ENV     = 'gnome'
                 DE_MAJ_VER      = '3'
```

Bare metal or virtual machine: 
(If in VM, which VM software): VMWare Workstation Pro 17.5.2 build-23775571

Keyboard type (IBM, Chromebook, Windows, Apple):  Windows (I guess) Logitech MX Keys Mini
Keyboard device name (try 'toshy-devices'): 

```
$ toshy-devices 

List of devices seen by the keymapper (xwaykeyz or keyszer): 

Neither "xwaykeyz" nor "keyszer" command was found in: 
/home/vmuser1/.config/toshy/.venv/bin:/home/vmuser1/.nvm/versions/node/v18.17.1/bin:/home/vmuser1/.local/bin:/home/vmuser1/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/vmuser1/.local/share/JetBrains/Toolbox/scripts:/home/vmuser1/.local/share/JetBrains/Toolbox/bin:/home/vmuser1/.local/share/JetBrains/Toolbox/scripts:/home/vmuser1/.local/share/JetBrains/Toolbox/scripts.
Toshy config cannot be loaded until one of these is installed.
```
* * *

Executed `setup_toshy.py prep-only`  as admin user. Completed successfully
Executed `setup_toshy.py install` as non admin user (who can still execute sudo). Failed See [toshy-out.txt]

**Comment by RedBearAK** [comment]
Isn’t this probably the same issue of needing to remove the older Python “devel” package that interferes? 

```sh
Downloading Packages:
(1/2): python36-devel-3.6.8-39.module+el8.10.0+20784+edafcd43.x86_64.rpm                                                      30 kB/s |  17 kB     00:00    
(2/2): platform-python-devel-3.6.8-62.el8_10.x86_64.rpm                                                                      295 kB/s | 241 kB     00:00    
```

The reason the reinstall was needed was most likely the new Python version. Looks like RHEL 8.10 brought in Python 3.12 to replace 3.11. 

I think once you do the same thing you did last time it should succeed in finishing.

**Comment by jeff9finger-usda** [comment]
Yes. I think that is correct. I mainly6 just wanted to report it, for visibility. I will try the things we tried before and let you know

**Comment by jeff9finger-usda** [comment]
Executed the following and it works. Hope this helps others. @RedBearAK Thanks

```
$ sudo dnf remove python36-devel
$ sudo dnf install python3.12-devel
$ ./setup_toshy.py install
```

**Comment by RedBearAK** [comment]
@jeff9finger-usda 

I appreciate the thorough report and log attachments, and the follow-up showing the same solution still works. I'll continue to try to find a better solution to the package conflict during the build process, when I get some time.

**Comment by RedBearAK** [comment]
@jeff9finger-usda 

I took a closer look at the way the installer handles the "quirks" of RHEL 8, to get it into a state to continue installing the full package list. As part of that, it figures out the latest version of the Python packages that are available to use for the Python virtual environment, to get something newer than the default Python 3.6 that isn't compatible with the keymapper. So it's already pre-installing a few packages like `python3.12-devel` on RHEL 8.10 (or AlmaLinux). 

However it was not installing `python3.12-pip`, which it probably should have been. And, the primary cause of your issue is most likely that I wasn't removing the `python3-devel` package name, so it somehow ends up also installing `python36-devel`. Now I'm having the quirks handler strip out those version-generic packages, so maybe that will help. 

I still don't know why this isn't a problem with AlmaLinux 8. It doesn't install `python36-devel` for some reason. 

I've tested it in AlmaLinux 8.10 just to make sure the install still works with the changes. Seems fine. I know you don't have much time to mess around with stuff like this, so it's fine if you can't get around to testing it. I'll just assume that it at least won't make anything worse. 

The only way to test whether the new changes will keep the problem from happening on your system is for you to test it yourself. And you'd have to do it in this sequence for it to mean anything: 

1. Test the installer you downloaded previously, starting from the `prep-only` step that installs packages, to see if the problem with the conflicting package recurs.
2. If the problem recurs, grab the new `dev_beta` zip.
3. Remove the conflicting package like you did before. (Important!)
4. Test the new installer from the `dev_beta` zip, starting from the `prep-only` step. 
5. Cross fingers.

I'll re-open this since technically it isn't confirmed as resolved yet. 

If you can't get to it before RHEL 8.11 arrives, or some other system maintenance happens, either keep hold of the previous installer you downloaded or you'd need to grab the zip from the v24.06 release to have equivalent code. I may end up merging this into main before long, so the main branch would change its behavior.

**Comment by RedBearAK** [comment]
@jeff9finger-usda 

I managed to find a couple of things in the core property matching function that I could change to reduce CPU usage while typing. One of the changes was pretty significant. It's in the `dev_beta` branch for now. After a bit more testing I'll probably merge into `main`. 

You may want to check it out since you're having to work in a virtual machine, with possibly limited resources compared to running on the host. 

Before this change it was possible to easily max out a thread (or a core on a CPU with no threads) by just rapidly typing. Now the keymapper will use no more than about 2-3% of my CPU in normal typing, on a 4c/8t Ryzen 3700u (fairly old by now). The maxed out thread was about 12.5% of the CPU. I can't get anywhere near that now, even while mashing keys much faster than most people will ever be able to type. 

With this change there should be less of a problem with a noticeable delay in typing when a system is really busy. Some users noticed that, including me at times. It may still happen with a really busy system, but I think it will be far less prevalent.

**Comment by RedBearAK** [comment]
@jeff9finger-usda 

I think I finally discovered the mistake I made that caused so much trouble with the Python version in the virtual environment, on RHEL. The problem that kept causing errors during the build of the `pip` packages. 

Turns out in the specific case of RHEL, the installer script was probably always installing the system Python version (3.9) in the virtual environment, not the newer Python version. Despite what the log said about the Python version that was being used to create the `venv`. The critical command used to actually create the `venv` was not being updated to the newer Python interpreter path. Now the command is computed dynamically when called, so it should always be correct. 

Changed from a simple class attribute to a class method with a `@property` decorator: 

```py
    @property
    def venv_cmd_lst(self):
        # Originally a class instance attribute variable:
        # self.venv_cmd_lst           = [self.py_interp_path, '-m', 'venv', self.venv_path]
        # Needs to re-evaluate itself when accessed, in case Python interpreter path changed:
        return [self.py_interp_path, '-m', 'venv', self.venv_path]
```

That's the way it always should have been. In the RHEL 8 case, this should correctly call on the `python3.12` binary to create the virtual environment. And I also fix some build-related paths to match that version of Python. 

I'm hoping all this means if you ever need to reinstall Toshy again on RHEL, it will just work correctly without any weird adjustments to `devel` packages. 🤞🏽 🙏🏽 

There were some performance optimizations in the Python 3.10, 3.11, and 3.12 releases, so an install from this newer state of the Toshy installer should also result in a bit more efficiency from the keymapper. Independent of anything I did on my side. Measurable, maybe, but I don't know how obvious it would be in regular usage.

**Comment by RedBearAK** [comment]
@jeff9finger-usda 

This is separate from the issue of your main user not having access to `sudo` during the Toshy install. In that case, the first install on a system like that will still require the `prep-only` step under an admin user. Usually that won't have to be repeated on the same system, unless there's some new package in the package list that wasn't installed previously. I don't think that should be the case. 

The fix I was talking about is more regarding the strange issue you kept having during the build of the Python `pip` packages like `dbus-python`, inside the Python virtual environment. Where it was looking for build-related files in the wrong place and not finding them, so the build failed with an error. I really hope that's not going to happen the next time around. I definitely fixed something that was not being done correctly. 

But if you do have issues again, post the details here, unless it seems to be a very different issue. 

Everything is in the main branch, so just download the zip from the main page when you need to. The "releases" as always are behind the current state of the main branch.

**Comment by jeff9finger-usda** [comment]
Was able to try again, but it doesn't work :-(

Here is the output:
```
$ ./setup_toshy.py install


(!!) NOTICE: It is ESSENTIAL to have your system completely updated.

Have you updated your system recently? [y/N]: y

§  Getting environment information...
================================================================================
The active init system is: 'systemd' (Systemd)

(EV) Toshy installer sees this environment:
	 DISTRO_ID        = 'rhel'
	 DISTRO_VER       = '8.10'
	 VARIANT_ID       = 'notfound'
	 SESSION_TYPE     = 'x11'
	 DESKTOP_ENV      = 'gnome'
	 DE_MAJ_VER       = '3'
	 WINDOW_MGR       = 'gnome-shell'

Is user "vmuser1" an admin user that can run "sudo" commands? [y/n]: n



ALERT!  ALERT!  ALERT!  ALERT!  ALERT!  ALERT!  ALERT!  ALERT!  ALERT!  ALERT!

The secret code for this run is "iLGS". You will need this.

It is possible to install as an unprivileged user, but only after an admin user
first runs the full install or a "prep-only" sequence. The admin user must
install from a full desktop session, or from a "su --login adminuser" shell
instance. The admin user can do just the "prep" steps with:

./setup_toshy.py prep-only

... instead of using:

./setup_toshy.py install

Use the "prep-only" command if it is not desired that Toshy should also run when
the admin user logs into a desktop session. When using "su --login adminuser",
that user will also need to download an independent copy of the Toshy zip file
to install from, using a "wget" or "curl" command. Or use "sudo" to copy the zip
file from the unprivileged user's Downloads folder. See the Wiki for a better
example of the full "prep-only" sequence with a separate admin user.

If you understand everything written above or already
took care of prepping the system and want to proceed
with an unprivileged install, enter the secret code: iLGS

Good code. Continuing with an unprivileged install of Toshy user components...


§  Cloning keymapper branch...
================================================================================
Cloning branch "main" from https://github.com/RedBearAK/xwaykeyz.git...
Cloning into '/tmp/toshy-main/keymapper-temp'...
remote: Enumerating objects: 3158, done.
remote: Counting objects: 100% (127/127), done.
remote: Compressing objects: 100% (51/51), done.
remote: Total 3158 (delta 80), reused 96 (delta 57), pack-reused 3031 (from 2)
Receiving objects: 100% (3158/3158), 695.35 KiB | 810.00 KiB/s, done.
Resolving deltas: 100% (2023/2023), done.
   >> Task completed successfully <<   


§  Backing up existing Toshy config folder...
================================================================================
Prepared existing config file data for merging into new config.
No existing preferences db file found in /home/vmuser1/.config/toshy.
Backup completed to '/home/vmuser1/.config/toshy_config_backups/toshy_20250205_122909'
   >> Task completed successfully <<   


§  Installing Toshy files...
================================================================================
Installed default Toshy config file.
Toshy files installed in '/home/vmuser1/.config/toshy'.
Existing user customizations applied to the new config file.
   >> Task completed successfully <<   


§  Setting up the Python virtual environment...
================================================================================
Handling Python virtual environment quirks in RHEL-type distros...
Using Python version 3.12.
Full venv command: /bin/python3.12 -m venv /home/vmuser1/.config/toshy/.venv
Python virtual environment setup complete.
Location: '/home/vmuser1/.config/toshy/.venv'
   >> Task completed successfully <<   


§  Installing/upgrading Python packages...
================================================================================
Setting PYTHONPATH to: /usr/include/python3.12
Setting CFLAGS to: -I/usr/include/python3.12
Setting LDFLAGS to: -L/usr/lib64
Requirement already satisfied: pip in /home/vmuser1/.config/toshy/.venv/lib64/python3.12/site-packages (23.2.1)
Collecting pip
  Obtaining dependency information for pip from https://files.pythonhosted.org/packages/85/8a/1ddf40be20103bcc605db840e9ade09c8e8c9f920a03e9cfe88eae97a058/pip-25.0-py3-none-any.whl.metadata
  Using cached pip-25.0-py3-none-any.whl.metadata (3.7 kB)
Using cached pip-25.0-py3-none-any.whl (1.8 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.2.1
    Uninstalling pip-23.2.1:
      Successfully uninstalled pip-23.2.1
Successfully installed pip-25.0
Collecting wheel
  Using cached wheel-0.45.1-py3-none-any.whl.metadata (2.3 kB)
Using cached wheel-0.45.1-py3-none-any.whl (72 kB)
Installing collected packages: wheel
Successfully installed wheel-0.45.1
Collecting setuptools
  Using cached setuptools-75.8.0-py3-none-any.whl.metadata (6.7 kB)
Using cached setuptools-75.8.0-py3-none-any.whl (1.2 MB)
Installing collected packages: setuptools
Successfully installed setuptools-75.8.0
Collecting pillow
  Using cached pillow-11.1.0-cp312-cp312-manylinux_2_28_x86_64.whl.metadata (9.1 kB)
Using cached pillow-11.1.0-cp312-cp312-manylinux_2_28_x86_64.whl (4.5 MB)
Installing collected packages: pillow
Successfully installed pillow-11.1.0
Collecting lockfile
  Using cached lockfile-0.12.2-py2.py3-none-any.whl.metadata (2.4 kB)
Collecting dbus-python
  Using cached dbus-python-1.3.2.tar.gz (605 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [101 lines of output]
      + meson setup /tmp/pip-install-sw1vr9z_/dbus-python_08a194009244485f9c30f2ee54713bf4 /tmp/pip-install-sw1vr9z_/dbus-python_08a194009244485f9c30f2ee54713bf4/.mesonpy-vnt7capk -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=/tmp/pip-install-sw1vr9z_/dbus-python_08a194009244485f9c30f2ee54713bf4/.mesonpy-vnt7capk/meson-python-native-file.ini
      The Meson build system
      Version: 1.7.0
      Source dir: /tmp/pip-install-sw1vr9z_/dbus-python_08a194009244485f9c30f2ee54713bf4
      Build dir: /tmp/pip-install-sw1vr9z_/dbus-python_08a194009244485f9c30f2ee54713bf4/.mesonpy-vnt7capk
      Build type: native build
      Project name: dbus-python
      Project version: 1.3.2
      C compiler for the host machine: cc (gcc 8.5.0 "cc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-22)")
      C linker for the host machine: cc ld.bfd 2.30-125
      Host machine cpu family: x86_64
      Host machine cpu: x86_64
      Compiler for C supports arguments -fno-common: YES
      Compiler for C supports arguments -Wno-missing-field-initializers: YES
      Compiler for C supports arguments -Wno-declaration-after-statement: YES
      Compiler for C supports arguments -Wno-inline: YES
      Compiler for C supports arguments -Wno-redundant-decls: YES
      Compiler for C supports arguments -Wno-switch-default: YES
      Compiler for C supports arguments -Wno-write-strings: YES
      Compiler for C supports arguments -Wcast-align: YES
      Compiler for C supports arguments -Wdouble-promotion: YES
      Compiler for C supports arguments -Wduplicated-cond: YES
      Compiler for C supports arguments -Wfloat-equal: YES
      Compiler for C supports arguments -Wformat-nonliteral: YES
      Compiler for C supports arguments -Wformat-security: YES
      Compiler for C supports arguments -Wformat=2: YES
      Compiler for C supports arguments -Winit-self: YES
      Compiler for C supports arguments -Wlogical-op: YES
      Compiler for C supports arguments -Wmissing-declarations: YES
      Compiler for C supports arguments -Wmissing-format-attribute: YES
      Compiler for C supports arguments -Wmissing-include-dirs: YES
      Compiler for C supports arguments -Wmissing-noreturn: YES
      Compiler for C supports arguments -Wnull-dereference: YES
      Compiler for C supports arguments -Wpacked: YES
      Compiler for C supports arguments -Wpointer-arith: YES
      Compiler for C supports arguments -Wshadow: YES
      Compiler for C supports arguments -Wswitch-enum: YES
      Compiler for C supports arguments -Wundef: YES
      Compiler for C supports arguments -Wunused-but-set-variable: YES
      Compiler for C supports arguments -Wjump-misses-init: YES
      Compiler for C supports arguments -Wmissing-prototypes: YES
      Compiler for C supports arguments -Wnested-externs: YES
      Compiler for C supports arguments -Wold-style-definition: YES
      Compiler for C supports arguments -Wpointer-sign: YES
      Compiler for C supports arguments -Wstrict-prototypes: YES
      Configuring _dbus-python-config.h using configuration
      
      Executing subproject dbus-gmain
      
      dbus-gmain| Project name: dbus-gmain
      dbus-gmain| Project version: undefined
      dbus-gmain| C compiler for the host machine: cc (gcc 8.5.0 "cc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-22)")
      dbus-gmain| C linker for the host machine: cc ld.bfd 2.30-125
      dbus-gmain| Compiler for C supports arguments -fno-common: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wno-missing-field-initializers: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wcast-align: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wdouble-promotion: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wduplicated-branches: YES
      dbus-gmain| Compiler for C supports arguments -Wduplicated-cond: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wfloat-equal: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wformat-nonliteral: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wformat-security: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wformat=2: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Winit-self: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wlogical-op: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wmissing-declarations: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wmissing-format-attribute: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wmissing-include-dirs: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wmissing-noreturn: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wnull-dereference: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wpacked: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wpointer-arith: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wredundant-decls: YES
      dbus-gmain| Compiler for C supports arguments -Wshadow: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wswitch-default: YES
      dbus-gmain| Compiler for C supports arguments -Wswitch-enum: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wundef: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wunused-but-set-variable: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wwrite-strings: YES
      dbus-gmain| Compiler for C supports arguments -Wdeclaration-after-statement: YES
      dbus-gmain| Compiler for C supports arguments -Wjump-misses-init: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wmissing-prototypes: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wnested-externs: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wold-style-definition: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wpointer-sign: YES (cached)
      dbus-gmain| Compiler for C supports arguments -Wstrict-prototypes: YES (cached)
      dbus-gmain| Found pkg-config: YES (/bin/pkg-config) 1.4.2
      dbus-gmain| Run-time dependency dbus-1 found: YES 1.12.8
      dbus-gmain| Run-time dependency glib-2.0 found: YES 2.56.4
      dbus-gmain| Run-time dependency gthread-2.0 found: YES 2.56.4
      dbus-gmain| Program dbus-run-session found: YES (/bin/dbus-run-session)
      dbus-gmain| Build targets in project: 1
      dbus-gmain| Subproject dbus-gmain finished.
      
      Program python found: YES (/home/vmuser1/.config/toshy/.venv/bin/python)
      Dependency dbus-1 found: YES 1.12.8 (cached)
      Run-time dependency python found: NO (tried pkgconfig, pkgconfig and sysconfig)
      
      ../meson.build:175:11: ERROR: Python dependency not found
      
      A full log can be found at /tmp/pip-install-sw1vr9z_/dbus-python_08a194009244485f9c30f2ee54713bf4/.mesonpy-vnt7capk/meson-logs/meson-log.txt
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
(EE) Problem installing/upgrading Python packages. Installer exiting.
```

**Comment by RedBearAK** [comment]
That's a bit disheartening. 

Does the old trick of removing the Python 3.6 devel package accomplish anything? It shouldn't have come back if you didn't do the `prep-only` step as an admin user before trying the non-admin reinstall, but... Maybe a system update brings it back? 

```
$ sudo dnf remove python36-devel
$ sudo dnf install python3.12-devel
$ ./setup_toshy.py install
```

There's still the `--no-dbus-python` option, but that reduces functionality a bit.

**Comment by jeff9finger-usda** [comment]
Thanks for the suggestion. This worked! 
`python36-devel` was not installed, so it appears that the key for me was installing `python3.12-devel`

Now, I'm trying to get the appindicator extension installed.

**Comment by RedBearAK** [comment]
>  python36-devel was not installed, so it appears that the key for me was installing python3.12-devel

Hmm, maybe you actually didn't do the `prep-only` package install step as the admin since the installer was smartened up enough to specifically install the `python3.12-devel` (or whatever the latest available Python version is) package. 

I still have high hopes that if you have to start over in a new, clean VM instance the process will work better. Thought you will definitely need to remember to do the `prep-only` step first in that case. 

> Now, I'm trying to get the appindicator extension installed.

Is there another error? If so, maybe you really do need to do the `prep-only` step again, as the admin user. That should clear the way for everything going into the Python virtual environment folder, as the regular user. 

Supposedly.

**Comment by RedBearAK** [comment]
> Now, I'm trying to get the appindicator extension installed.

Oh, not the `libappindicator` package, the GNOME extension. Never mind. 😆

**Comment by jeff9finger-usda** [comment]
FYI: I did execute the `prep-only` option as an admin user. Not sure what happened here. But I will say that there is probably a strange and non standard configuration as the government beaurocrats seem to use a hammer to enforce what they think is security, even in the VM guest OS.

**Comment by RedBearAK** [comment]
That makes the mystery even deeper. If you downloaded the same new zip file in both accounts and did the `prep-only` step, that Python 3.12 package should have been installed. The fact that you can install the package easily in the non-admin user account with `sudo` to get past the remaining issue becomes quite odd. 

Well, we can agree that there is something non-standard going on in your VM, since I've never had trouble with any of this in the RHEL clones. I have no real choice but to close this up again, unless I get some other idea someday about what might be going wrong. 

As long as the Python version inside the `venv` is now Python 3.12 like it should be, things have at least improved in some way. 

```
[testuser@localhost ~]$ python3 --version
Python 3.6.8
[testuser@localhost ~]$ source toshy-venv

 To activate the Toshy Python virtual environment, run this command: 

			 source toshy-venv 

 (Ignore the text above if you've already run the command as shown.) 

(.venv) [testuser@localhost ~]$ python3 --version
Python 3.12.8
```

**Comment by jeff9finger-usda** [comment]
OK. Thanks for working on this.
